command/action.go

```diff
@@ -128,189 +123,39 @@ func (c *ActionCommand) Run(args []string) int {
-		return 1
-	}
-
-	client, err := c.Meta.Client()
-	if err != nil {
-		c.Ui.Error(fmt.Sprintf("Error initializing client: %v", err))
-		return 1
-	}
-
-	var allocStub *api.AllocationListStub
-	// If no allocation provided, grab a random one from the job
-	if allocation == "" {
-
-		// Group param cannot be empty if allocation is empty,
-		// since we'll need to get a random allocation from the group
-		if group == "" {
-			c.Ui.Error("A group name is required if no allocation is provided")
-			c.Ui.Error(commandErrorText(c))
-			return 1
-		}
-
-		if task == "" {
-			c.Ui.Error("A task name is required if no allocation is provided")
-			c.Ui.Error(commandErrorText(c))
-			return 1
-		}
-
-		jobID, ns, err := c.JobIDByPrefix(client, job, nil)
-		if err != nil {
-			c.Ui.Error(err.Error())
-			return 1
-		}
-
-		allocStub, err = getRandomJobAlloc(client, jobID, group, ns)
-		if err != nil {
-			c.Ui.Error(fmt.Sprintf("Error fetching allocations: %v", err))
-			return 1
-		}
-	} else {
-		allocs, _, err := client.Allocations().PrefixList(sanitizeUUIDPrefix(allocation))
-		if err != nil {
-			c.Ui.Error(fmt.Sprintf("Error querying allocation: %v", err))
-			return 1
-		}
-
-		if len(allocs) == 0 {
-			c.Ui.Error(fmt.Sprintf("No allocation(s) with prefix or id %q found", allocation))
-			return 1
-		}
-
-		if len(allocs) > 1 {
-			out := formatAllocListStubs(allocs, false, shortId)
-			c.Ui.Error(fmt.Sprintf("Prefix matched multiple allocations\n\n%s", out))
-			return 1
-		}
-
-		allocStub = allocs[0]
-	}
-
-	q := &api.QueryOptions{Namespace: allocStub.Namespace}
-	alloc, _, err := client.Allocations().Info(allocStub.ID, q)
-	if err != nil {
-		c.Ui.Error(fmt.Sprintf("Error querying allocation: %s", err))
-		return 1
-	}
+	var commandWithFlags []string
 
 	if task != "" {
-		err = validateTaskExistsInAllocation(task, alloc)
-	} else {
-		task, err = lookupAllocTask(alloc)
+		commandWithFlags = append(commandWithFlags, "-task="+task)
 	}
-	if err != nil {
-		c.Ui.Error(err.Error())
-		return 1
+	if group != "" {
+		commandWithFlags = append(commandWithFlags, "-group="+group)
 	}
-
-	if !stdinOpt {
-		c.Stdin = bytes.NewReader(nil)
+	if allocation != "" {
+		commandWithFlags = append(commandWithFlags, "-alloc="+allocation)
 	}
-
-	if c.Stdin == nil {
-		c.Stdin = os.Stdin
+	if job != "" {
+		commandWithFlags = append(commandWithFlags, "-job="+job)
 	}
 
-	if c.Stdout == nil {
-		c.Stdout = os.Stdout
+	if !stdinOpt {
+		commandWithFlags = append(commandWithFlags, "-i=false")
 	}
-
-	if c.Stderr == nil {
-		c.Stderr = os.Stderr
+	if !ttyOpt {
+		commandWithFlags = append(commandWithFlags, "-t=false")
 	}
-
-	action := args[0]
-
-	code, err := c.execImpl(client, alloc, task, job, action, ttyOpt, escapeChar, c.Stdin, c.Stdout, c.Stderr)
-	if err != nil {
-		c.Ui.Error(fmt.Sprintf("failed to exec into task: %v", err))
-		return 1
+	if escapeChar != defaultEscapeChar {
+		commandWithFlags = append(commandWithFlags, "-e="+escapeChar)
```

Comment:
Q: I'm only passing flags/args if they're non-default. Is this the right way to do it? I'm trying to reason about how "upstream" we want our defaults to live.

Reply:
I think you probably want to pass the args in directly w/o messing with them. That way each downstream command implementation can deal with them on its own. That's how `nomad status` does it.